python/shark_turbine/aot/support/procedural/primitives.py

```diff
@@ -58,6 +63,33 @@ class IrScalar(Intrinsic):
     def __init__(self, ir_type: IrType):
         self.ir_type = ir_type
 
+    def __add__(self, other):
+        t = current_ir_trace()
+        with t.ip, t.loc:
+            # Type check and promotion.
+            # TODO: Add more comprehensive type promotion hiearchy as seen in
+            # https://jax.readthedocs.io/en/latest/jep/9407-type-promotion.html
+            lhs = self.ir_value
+            if isinstance(other, IrScalar):
+                # Assumes when both are Value, they have same type.
+                rhs = other.ir_value
+            elif isinstance(other, (int, bool)):
+                rhs = arith_d.ConstantOp(lhs.type, other).result
+            elif isinstance(other, float) and _is_integer_like_type(self.ir_type):
+                lhs = arith_d.SIToFPOp(F32Type.get(), lhs).result
+                rhs = arith_d.ConstantOp(F32Type.get(), other).result
+            assert lhs.type == rhs.type, "Mismatch type between lhs and rhs."
```

Comment:
Generally, should raise a ValueError for "user problems" and reserve asserts for some kind of internal inconsistency.

Reply:
done!